File: website/vo2-cs/Gpx.cs
using System.Globalization;

namespace Vo2;

internal static class Gpx
{
    private const int MaxTrackPoints = 200_000;

    private static bool TryParseAttrDouble(string tag, string attr, out double value)
    {
        value = 0;
        var pat = attr + "=\"";
        var idx = tag.IndexOf(pat, StringComparison.Ordinal);
        if (idx < 0) return false;
        var rest = tag.AsSpan(idx + pat.Length);
        var end = rest.IndexOf('"');
        if (end < 0) return false;
        return double.TryParse(rest[..end], NumberStyles.Float, CultureInfo.InvariantCulture, out value);
    }

    private static ReadOnlySpan<char> LocalTagName(string tag)
    {
        if (tag.Length == 0) return ReadOnlySpan<char>.Empty;
        switch (tag[0])
        {
            case '/': case '!': case '?': return ReadOnlySpan<char>.Empty;
        }
        var span = tag.AsSpan();
        var ws = span.IndexOfAny([' ', '\t', '\r', '\n']);
        if (ws >= 0) span = span[..ws];
        span = span.TrimEnd('/');
        var colon = span.LastIndexOf(':');
        if (colon >= 0) span = span[(colon + 1)..];
        return span;
    }

    /// <summary>
    /// Byte-driven, namespace-tolerant GPX parser. Extracts trkpt elements
    /// and their ele/time/hr children; tolerates CDATA, comments, and
    /// arbitrary namespace prefixes. Mirrors the Rust implementation.
    /// </summary>
    public static List<TrackPoint> Parse(string content)
    {
        var points = new List<TrackPoint>(1024);
        TrackPoint? current = null;

        var i = 0;
        while (i < content.Length)
        {
            if (content[i] != '<') { i++; continue; }
            var tagStart = i + 1;

            // CDATA
            if (tagStart + 8 <= content.Length
                && content.AsSpan(tagStart, 8).SequenceEqual("![CDATA["))
            {
                var off = content.IndexOf("]]>", i, StringComparison.Ordinal);
                if (off < 0) throw new FormatException("Unterminated CDATA section");
                i = off + 3;
                continue;
            }
            // Comment
            if (tagStart + 3 <= content.Length
                && content.AsSpan(tagStart, 3).SequenceEqual("!--"))
            {
                var off = content.IndexOf("-->", i, StringComparison.Ordinal);
                if (off < 0) throw new FormatException("Unterminated comment");
                i = off + 3;
                continue;
            }

            var j = tagStart;
            while (j < content.Length && content[j] != '>') j++;
            if (j >= content.Length) break;
            var tag = content.Substring(tagStart, j - tagStart);
            i = j + 1;

            var local = LocalTagName(tag);

            if (local.SequenceEqual("trkpt"))
            {
                if (TryParseAttrDouble(tag, "lat", out var lat)
                    && TryParseAttrDouble(tag, "lon", out var lon))
                {
                    current = new TrackPoint { Lat = lat, Lon = lon };
                }
            }
            else if (tag.StartsWith("/trkpt", StringComparison.Ordinal))
            {
                if (current != null)
                {
                    points.Add(current);
                    current = null;
                    if (points.Count > MaxTrackPoints)
                    {
                        throw new FormatException(
                            $"GPX file exceeds {MaxTrackPoints} track points — file is too large to analyse");
                    }
                }
            }
            else if (current != null && local.Length > 0)
            {
                if (local.SequenceEqual("ele") || local.SequenceEqual("time") || local.SequenceEqual("hr"))
                {
                    var textStart = i;
                    while (i < content.Length && content[i] != '<') i++;
                    var text = content[textStart..i].Trim();
                    if (text.Length == 0) continue;

                    if (local.SequenceEqual("ele"))
                    {
                        if (double.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out var ele))
                            current.Ele = ele;
                    }
                    else if (local.SequenceEqual("time"))
                    {
                        if (Timestamp.TryParse(text, out var t)) current.TimeS = t;
                    }
                    else // hr
                    {
                        if (uint.TryParse(text, NumberStyles.Integer, CultureInfo.InvariantCulture, out var hr))
                            current.Hr = hr;
                    }
                }
            }
        }

        return points;
    }
}

File: website/vo2/as/assembly/gpx.ts
import { parseTimestamp } from './timestamp';

const MAX_TRACK_POINTS: i32 = 200_000;

// Track points use NaN for missing ele/time, and -1 for missing hr.
export class TrackPoint {
  lat: f64 = 0;
  lon: f64 = 0;
  ele: f64 = NaN;
  timeS: f64 = NaN;
  hr: i32 = -1;
}

function parseAttrFloat(tag: string, attr: string): f64 {
  const pat = attr + '="';
  const idx = tag.indexOf(pat);
  if (idx < 0) return NaN;
  const rest = tag.substring(idx + pat.length);
  const end = rest.indexOf('"');
  if (end < 0) return NaN;
  return parseFloat(rest.substring(0, end));
}

function localTagName(tag: string): string {
  if (tag.length == 0) return '';
  const c = tag.charCodeAt(0);
  if (c == 47 /* '/' */ || c == 33 /* '!' */ || c == 63 /* '?' */) return '';
  let name = tag;
  // Strip everything after first whitespace
  let ws = -1;
  for (let i = 0; i < name.length; i++) {
    const ch = name.charCodeAt(i);
    if (ch == 32 || ch == 9 || ch == 10 || ch == 13) { ws = i; break; }
  }
  if (ws >= 0) name = name.substring(0, ws);
  while (name.endsWith('/')) name = name.substring(0, name.length - 1);
  const colon = name.lastIndexOf(':');
  if (colon >= 0) name = name.substring(colon + 1);
  return name;
}

export class ParseResult {
  points: TrackPoint[];
  error: string | null;
  constructor(points: TrackPoint[], error: string | null) {
    this.points = points; this.error = error;
  }
}

export function parseGpx(content: string): ParseResult {
  const points: TrackPoint[] = [];
  let current: TrackPoint | null = null;

  let i: i32 = 0;
  const n: i32 = content.length;
  while (i < n) {
    if (content.charCodeAt(i) != 60 /* '<' */) { i++; continue; }
    const tagStart = i + 1;

    // CDATA: skip to ]]>
    if (content.startsWith('![CDATA[', tagStart)) {
      const off = content.indexOf(']]>', i);
      if (off < 0) return new ParseResult([], 'Unterminated CDATA section');
      i = off + 3;
      continue;
    }
    // Comment: skip to -->
    if (content.startsWith('!--', tagStart)) {
      const off = content.indexOf('-->', i);
      if (off < 0) return new ParseResult([], 'Unterminated comment');
      i = off + 3;
      continue;
    }

    let j = tagStart;
    while (j < n && content.charCodeAt(j) != 62 /* '>' */) j++;
    if (j >= n) break;
    const tag = content.substring(tagStart, j);
    i = j + 1;

    const local = localTagName(tag);

    if (local == 'trkpt') {
      const lat = parseAttrFloat(tag, 'lat');
      const lon = parseAttrFloat(tag, 'lon');
      if (!isNaN(lat) && !isNaN(lon)) {
        const pt = new TrackPoint();
        pt.lat = lat;
        pt.lon = lon;
        current = pt;
      }
    } else if (tag.startsWith('/trkpt')) {
      if (current != null) {
        points.push(current!);
        current = null;
        if (points.length > MAX_TRACK_POINTS) {
          return new ParseResult([],
            'GPX file exceeds ' + MAX_TRACK_POINTS.toString() + ' track points — file is too large to analyse');
        }
      }
    } else if (current != null && local.length > 0) {
      if (local == 'ele' || local == 'time' || local == 'hr') {
        const textStart = i;
        while (i < n && content.charCodeAt(i) != 60) i++;
        const text = content.substring(textStart, i).trim();
        if (text.length == 0) continue;
        if (local == 'ele') {
          const v = parseFloat(text);
          if (!isNaN(v)) current!.ele = v;
        } else if (local == 'time') {
          const v = parseTimestamp(text);
          if (!isNaN(v)) current!.timeS = v;
        } else {
          const v = <i32>parseInt(text, 10);
          if (v >= 0) current!.hr = v;
        }
      }
    }
  }

  return new ParseResult(points, null);
}
